### ISSUE TITLE
[$500]  LHN - Announce room on invited workspace member is not bold

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#35122]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.31-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. User A create workspace
2. User A navigates to announce room> settings> set welcome message
3. User A goes to Settings> Workspace> Members> Invite User B
4. Login as User B

## Expected Result:
Announce room on invited workspace member should be bold on LHN

## Actual Result:
Announce room of invited workspace member is not bold on LHN, even green line on welcome message appears
Also, welcome message does not change on employee side when admin change the content of welcome message

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/cb11f485-082b-41b5-874e-a5d31b2138ce

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016aee40f12650c4a6</li>
        <li>Upwork Job ID: 1750329685857648640</li>
        <li>Last Price Increase: 2024-02-15</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Contributor | 28129326</li>
</ul></li>
    </ul>
</details>

**Comment by jeremy-croff** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Announced messages to a user that gets invited - does not show up as unread messages in LHN. They do get a notification.

### What is the root cause of that problem?

<img width="439" alt="Screenshot 2024-01-24 at 8 55 15 PM" src="https://github.com/Expensify/App/assets/157416545/0fcc00a1-5af0-47fc-b840-3b29fc98b21b">
We currently create new reports with the same timestamp for lastRead and lastVisibleActionCreated

We only show something as unread if the last read time is before the last visible is created here:
https://github.com/Expensify/App/blob/0d0b0a8c9b073c30a9b54de15066789d17f7c718/src/libs/ReportUtils.ts#L3517

And the announced message, is merely metadata on the report. Not an actual message.


### What changes do you think we should make in order to solve the problem?
We should actually show all new channels as unread - this improves the customer experience by showing there's something new. Even if there are no messages in that channel.

We can make a small change here https://github.com/Expensify/App/blob/0d0b0a8c9b073c30a9b54de15066789d17f7c718/src/libs/ReportUtils.ts#L3517 to also be unread if the lastRead equals lastVisibleActionCreated, indicating the newly created report condition.

Once a user actually clicks on the report - it will be marked as read as the lastRead timestamp updates:


https://github.com/Expensify/App/assets/157416545/b615bf89-7f34-4fe6-86fa-3579a5638c0f

This also covers the announce use case in this issue.

### What alternative solutions did you explore? (Optional)

**Comment by dragnoir** [comment]
I think it should be fixed here https://github.com/Expensify/App/issues/31748

**Comment by Santhosh-Sellavel** [comment]
> I think it should be fixed here #31748

@dragnoir Can you elaborate on why? 
Is that regression from #31748?

**Comment by dragnoir** [comment]
@Santhosh-Sellavel check this https://github.com/Expensify/App/issues/33881
it's the same issue here and the same problem. The suggested solution is the same in boath https://github.com/Expensify/App/issues/33881 and https://github.com/Expensify/App/issues/31748 but the issue is always there. The solution is not complete and need to handle all the UNREAD cases.

PR update: https://github.com/Expensify/App/pull/34551/files
Second issue proposal: https://github.com/Expensify/App/issues/33881#issuecomment-1890980793

c+ note from the same issue https://github.com/Expensify/App/issues/33881#issuecomment-1891261939 and https://github.com/Expensify/App/issues/33881#issuecomment-1891280250

but the PR submitted there https://github.com/Expensify/App/pull/34551 is not tested well.

**Comment by Santhosh-Sellavel** [comment]
Waiting for feedback https://github.com/Expensify/App/issues/31748#issuecomment-1910790219

I'm unavailable next week, Please assign a new C+ Issue here if required while I am away, thanks!

https://github.com/Expensify/App/issues/31748#issuecomment-191079021

**Comment by lschurr** [comment]
Asking for a volunteer in Slack - https://expensify.slack.com/archives/C01GTK53T8Q/p1706555426712459

**Comment by lschurr** [comment]
Assigning to @situchan - could you review?

**Comment by situchan** [comment]
Seems like 2 bugs stated in OP. Checking reproducibility first

**Comment by situchan** [comment]
> Announce room of invited workspace member is not bold on LHN, even green line on welcome message appears

reproduced

<img width="1790" alt="Screenshot 2024-01-30 at 1 26 41 AM" src="https://github.com/Expensify/App/assets/108292595/83ef4880-4335-4210-898c-351e81103528">

----

> Also, welcome message does not change on employee side when admin change the content of welcome message

not reproduced

**Comment by situchan** [comment]
@jeremy-croff can you please explain the root cause in detail?

This has 2 bugs: (might have the same root cause though)

- why LHN text is not bolded when login first time after getting invited?
- why unread marker keeps showing even after read and revisit report?


https://github.com/Expensify/App/assets/108292595/284f38d4-e75c-4c06-aa1e-bda622477b15

**Comment by jeremy-croff** [comment]
> [App/src/libs/ReportUtils.ts](https://github.com/Expensify/App/blob/0d0b0a8c9b073c30a9b54de15066789d17f7c718/src/libs/ReportUtils.ts#L3517)

While both are adressed with my fix,

 I will report back soon why the unread marker keeps showing in current main.
 It has to do with the logic that determines unread within the report is seperate from the LHN.

**Comment by jeremy-croff** [comment]
> @jeremy-croff can you please explain the root cause in detail?
> 
> This has 2 bugs: (might have the same root cause though)
> 
> * why LHN text is not bolded when login first time after getting invited?
> * why unread marker keeps showing even after read and revisit report?
> 
>  Screen.Recording.2024-01-30.at.1.35.12.AM.mov

While both are adressed with my fix,

I will report back soon why the unread marker keeps showing in current main.
It has to do with the logic that determines unread within the report is seperate from the LHN.

**Comment by jeremy-croff** [comment]
> @jeremy-croff can you please explain the root cause in detail?
> 
> This has 2 bugs: (might have the same root cause though)


> * why LHN text is not bolded when login first time after getting invited?
> * why unread marker keeps showing even after read and revisit report?
We grab the last read time from the report https://github.com/Expensify/App/blob/e2ddbcb62a7504aacc886413c5bc8a3f6ca3f138/src/pages/home/report/ReportActionsList.js#L235
And compare the report actions against that timestamp for the unread marker in the chat 
https://github.com/Expensify/App/blob/e2ddbcb62a7504aacc886413c5bc8a3f6ca3f138/src/pages/home/report/ReportActionsList.js#L369
Then marks it as unread here based on the action timestamp and the report last read time stamp:
https://github.com/Expensify/App/blob/e2ddbcb62a7504aacc886413c5bc8a3f6ca3f138/src/pages/home/report/ReportActionsList.js#L369
the welcome message has a later timestamp.

Following screenshots show this: 
![Screenshot 2024-01-29 at 4 19 15 PM](https://github.com/Expensify/App/assets/157416545/80651cdf-295c-4f38-916f-95035df0d4c8)


This flow is different from the LHN unread. 
https://github.com/Expensify/App/blob/0d0b0a8c9b073c30a9b54de15066789d17f7c718/src/libs/ReportUtils.ts#L3517
Where we just compare the timestamps from the report. 
And as demonstrated are the same in this scenario.

**Comment by lschurr** [comment]
What's the latest on this one @situchan?

**Comment by situchan** [comment]
I am not following @jeremy-croff's solution

**Comment by jeremy-croff** [comment]
I was trying to go through this flow again, to attempt to provide a clearer explanation. But it seems a side effect or bug is not allowing us to set a default announce message anymore. 

![Screenshot 2024-02-05 at 10 51 26 PM](https://github.com/Expensify/App/assets/157416545/34403aa0-7f6c-4e92-87fb-b339d1a09275)

**Comment by jeremy-croff** [comment]
> @jeremy-croff can you please explain the root cause in detail?
> 
> This has 2 bugs: (might have the same root cause though)
> 
> * why LHN text is not bolded when login first time after getting invited?
> * why unread marker keeps showing even after read and revisit report?
> 
>  Screen.Recording.2024-01-30.at.1.35.12.AM.mov

@situchan
So what I am observing with the behaviour you mentioned,  is that the timestamps of lastVisibleActionCreated  and lastReadTime are identical upon initial workspace creation. Because they are the same timestamp it fails this check [unread]https://github.com/Expensify/App/blob/500dd0949123ae1cd7b84211c0093f3c832d6775/src/libs/ReportUtils.ts#L3562  and considers it read. That's why new channels with no new actions are not bolded. As demonstrated in this ticket.

The small change of updating the Unread logic to `return lastReadTime <= lastVisibleActionCreated;` here https://github.com/Expensify/App/blob/500dd0949123ae1cd7b84211c0093f3c832d6775/src/libs/ReportUtils.ts#L3562
Addresses both of your points, because subsequently a read api call will be made if you click on an unread channel with a welcome message. Because this was not happening, the unread markers were not getting cleared. If it starts happening they will be.


I think reviewing my proposal, I would say it could be a risk if lastReadTime is actually genuinely the same timestamp lastVisibleActionCreated during a chat, even though super unlikely. However, we can enhance the logic in a different way:
return lastReadTime < lastVisibleActionCreated || lastReadTime === created
Which will also fix this ticket.

**Comment by lschurr** [comment]
@situchan could you review?

**Comment by situchan** [comment]
Thanks. I will provide feedback tomorrow

**Comment by situchan** [comment]
Is this still reproducible?

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

LHN - Announce room on invited workspace member is not bold.

### What is the root cause of that problem?

The user journey flow of this issue is as follows

1. From user A, invite user B to a Workspace
2. Backend receives request from user A and creates a new report with two actions for user B
   1. The first action is a `CREATED` action to let user B join the workspace
   2. The second action is a `ADDCOMMENT` action which is a whisper message
3. User B receives update from the backend and there're two issues
   1. The Announce room is not bold
   2. The unread marker is remain after reading and revisiting the room

The root cause of this issue is that the new Announce room report of user B has same `lastReadTime` and `lastVisibleActionCreated` which are equal to the `created` time of the first `CREATED` action. I guessed it's because the second action is a whisper action which is not visible in LHN. 

| Report                                                       | ReportActions                                                |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| <img width="354" alt="Screenshot 2024-02-14 at 2 22 41 PM" src="https://github.com/Expensify/App/assets/117511920/0c3a4315-6bc9-4592-9385-273347daef91"> | <img width="357" alt="Screenshot 2024-02-14 at 2 52 50 PM" src="https://github.com/Expensify/App/assets/117511920/767be935-5ea4-44ad-9e7e-587bd8c5ed8d"> |

So, method [ReportUtils.isUnread](https://github.com/Expensify/App/blob/023d62cba900f4333f73c52eb735402a994cbc4b/src/libs/ReportUtils.ts#L3629) returns which results in the two issues as mentioned above

1. Announce room report in LHN is not bold

   https://github.com/Expensify/App/blob/023d62cba900f4333f73c52eb735402a994cbc4b/src/libs/SidebarUtils.ts#L305

2. The unread marker is not removed after read

​	https://github.com/Expensify/App/blob/023d62cba900f4333f73c52eb735402a994cbc4b/src/pages/home/report/ReportActionsList.js#L211-L214

### What changes do you think we should make in order to solve the problem?

We can fix this issue from frontend by comparing `report.lastReadTime` with the created time of last displayable action of the report. The last displayable action here includes whisper actions while last visible action currently refers to the last action that can be visible in LHN preview, see [shouldReportActionBeVisibleAsLastAction](https://github.com/Expensify/App/blob/023d62cba900f4333f73c52eb735402a994cbc4b/src/libs/ReportActionsUtils.ts#L424).

We can add new method in ReportActionsUtils.ts

```typescript
function getLastDisplayableActionCreatedTime(reportID: string): string {
    const reportActions = allReportActions?.[reportID] ?? {};
    const filteredReportActions = Object.entries(reportActions ?? {})
        .filter(([key, reportAction]) => shouldReportActionBeVisible(reportAction, key))
        .map((entry) => entry[1]);
    const sortedReportActions = getSortedReportActions(filteredReportActions, true);
    if (sortedReportActions.length === 0) {
        return '';
    }
    return sortedReportActions[0].created ?? '';
}
```

and this line of method `ReportUtils.isUnread`

https://github.com/Expensify/App/blob/023d62cba900f4333f73c52eb735402a994cbc4b/src/libs/ReportUtils.ts#L3635

with

```js
const lastVisibleActionCreated = ReportActionsUtils.getLastDisplayableActionCreatedTime(report.reportID);
```

See demo

https://github.com/Expensify/App/assets/117511920/f80b0924-4835-4d44-b06a-6207d32032e7

### What alternative solutions did you explore? (Optional)

We can also get this issue fix from backend. We can fix the edge case when the lastest dispalyable action is whisper action and set `report.lastVisibleActionCreated` equal to the created time of the whisper action. That said, `report.lastVisibleActionCreated` should be equal to the last *displayable* action in `ReportActionsView`.

**Comment by eh2077** [comment]
> Is this still reproducible?

@situchan Yes, this issue is still reproducible. Can you kindly also have a look at my proposal ^? Thanks!

**Comment by lschurr** [comment]
@situchan - Could you take a look at this one?

**Comment by situchan** [comment]
I am not able to see "Welcome message" option anymore in announce room settings after #34150 is merged.
I think OP and title needs to be updated.

**Comment by lschurr** [comment]
Thanks @situchan - So is this still a bug? What needs to happen to move it forward?

**Comment by eh2077** [comment]
@lschurr I think we still need to fix this issue as @situchan pointed out in the above https://github.com/Expensify/App/issues/35122#issuecomment-1915425921. We can just omit bug reproduction steps about the `Welcome message` feature which has been removed.



> This has 2 bugs: (might have the same root cause though)
> 
> * why LHN text is not bolded when login first time after getting invited?
> * why unread marker keeps showing even after read and revisit report?
> 
>  [Screen.Recording.2024-01-30.at.1.35.12.AM.mov](https://private-user-images.githubusercontent.com/108292595/300583039-284f38d4-e75c-4c06-aa1e-bda622477b15.mov?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0Mjk3NzYsIm5iZiI6MTcwODQyOTQ3NiwicGF0aCI6Ii8xMDgyOTI1OTUvMzAwNTgzMDM5LTI4NGYzOGQ0LWU3NWMtNGMwNi1hYTFlLWJkYTYyMjQ3N2IxNS5tb3Y_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMFQxMTQ0MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNDcwMGIyYmM1YzBhODUzODA5NTEwZWNlOWU0NDRhODJhNDZhMTdkY2IwZjkxYzliN2M4MjYxMzBhNDk4ZDI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.HorliuFgZKTswmNps7KjKdhzJS39W-LemRW_SqnozgI)

**Comment by eh2077** [comment]
@situchan  May I have your feedback on my proposal https://github.com/Expensify/App/issues/35122#issuecomment-1943227866? Thanks.

**Comment by lschurr** [comment]
Bump on this one @situchan

**Comment by situchan** [comment]
updating in a few hrs

**Comment by situchan** [comment]
Ideally, this should be fixed in backend.
So `report.lastVisibleActionCreated` should be `whisperAction.created` when last visible action is whisperAction.
Currently, whisper actions are not considered as visible action in backend when calculate `report.lastVisibleActionCreated`.

If we wanna do frontend fix, @eh2077's [proposal](https://github.com/Expensify/App/issues/35122#issuecomment-1943227866) looks good.

🎀 👀 🎀 C+ reviewed

**Comment by situchan** [comment]
I think this bug came from the fact that whisper action should not be visible in LHN as last message.

**Comment by thienlnam** [comment]
Ah interesting - thanks for the summary. We should fix this in the BE - though the query for determining visible report actions is quite complex. Not an urgent bug so going to change the priority

**Comment by lschurr** [comment]
Any update here @thienlnam?

**Comment by thienlnam** [comment]
I'm not currently working on this issue over other vip-vsb items, perhaps we can tie it to a wave and see if anyone is interested?

**Comment by lschurr** [comment]
Ok, got it. I'll review this later today.

**Comment by lschurr** [comment]
Asking for a volunteer in Slack here: https://expensify.slack.com/archives/C066HJM2CAZ/p1712079432372279

**Comment by lschurr** [comment]
Haven't heard anything in the Slack thread - ideas on this one @thienlnam?

**Comment by thienlnam** [comment]
Nope, hasn't been a focus - I think it's fine to leave unassigned or even close for other priorities at this time

**Comment by lschurr** [comment]
Going to close this as not a priority for now.
